Makefile.cache

```diff
@@ -0,0 +1,637 @@
+	# 1. First load from Local cache path
+	# 2. If not, load from global cache path
+	$(eval CACHE_FILE_SELECT:=$(or $(wildcard $(SONIC_DPKG_LOCAL_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)), \
+                                   $(wildcard $(SONIC_DPKG_CACHE_DIR)/$($(1)_MOD_CACHE_FILE))) )
+
+	# Check if any of the derived package is not built
+	$(eval LOAD_DRV_DEB := $(foreach pkg,$(addprefix $($(1)_DST_PATH)/,$(1) $($(1)_DERIVED_DEBS) $($(1)_EXTRA_DEBS)),$(if $(wildcard $(pkg)),,$(pkg))))
+
+	# Load the cache if cache is eanbled and cache file is present in the cache
+	# Update the cache_loaded variable
+	$(if $(and $(CACHE_FILE_SELECT),$(filter cache,$(SONIC_DPKG_CACHE_METHOD))),
+		$(if $(LOAD_DRV_DEB), $($(1)_CACHE_USER) tar -C $($(1)_BASE_PATH) -mxzvf $(CACHE_FILE_SELECT) 1>> $($(1)_DST_PATH)/$(1).log ,echo );
+		echo "File $(CACHE_FILE_SELECT) is loaded from cache" >> $($(1)_DST_PATH)/$(1).log
+		$(eval $(1)_CACHE_LOADED := Yes)
+		$(shell touch $(CACHE_FILE_SELECT))
+		echo "[ CACHE::LOADED ] $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)" >> $($(1)_DST_PATH)/$(1).log
+		,
+		echo "File $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)  is not present in cache or cache mode set as $(SONIC_DPKG_CACHE_METHOD) !" >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] $($(1)_CACHE_DIR)/$($(1)_MOD_CACHE_FILE)" >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] DEP_FILES - Modified Files: [$($(1)_FILES_MODIFIED)] " >> $($(1)_DST_PATH)/$(1).log
+		echo "[ CACHE::SKIPPED ] DEPENDS   - Modified Files: [$?] " >> $($(1)_DST_PATH)/$(1).log
+	 )
+endef
+
+# Saves the deb package into debian cache
+# A single tared-zip cache is created for .deb and its derived packages in the cache direcory.
+# It saves the .deb into global cache only when its dependencies are not changed,
+#    Otherwise it saves the .deb into local cache
+# The cache save is protected with lock.
+# The SAVE_INTO_CACHE macro has dependecy with LOAD_FROM_CACHE macro
+#      The target specific variables -_MOD_SRC_PATH, _MOD_CACHE_FILE and _FILES_MODIFIED are
+#      derived from the LOAD_FROM_CACHE macro
+# Args:
+# $(1) => target name
+# $(2) => target output file name
+define SAVE_INTO_CACHE
+	$(eval MOD_CACHE_FILE=$($(1)_MOD_CACHE_FILE))
+	$(call MOD_LOCK,$(1),$(SONIC_DPKG_CACHE_DIR),$(MOD_CACHE_LOCK_SUFFIX),$(MOD_CACHE_LOCK_TIMEOUT))
+	$(if $($(1)_FILES_MODIFIED),
+		echo "Target $(1) dependencies are modifed - global save cache skipped" >> $($(1)_DST_PATH)/$(1).log
+		$(eval $(1)_CACHE_DIR := $(SONIC_DPKG_LOCAL_CACHE_DIR))
+	 )
+	$($(1)_CACHE_USER) tar -C $($(1)_BASE_PATH) -mczvf $($(1)_CACHE_DIR)/$(MOD_CACHE_FILE) $(2) $(addprefix $($(1)_DST_PATH)/,$($(1)_DERIVED_DEBS) $($(1)_EXTRA_DEBS) )  \
+        1>>$($(1)_DST_PATH)/$(1).log
+	sudo chmod 777 $($(1)_CACHE_DIR)/$(MOD_CACHE_FILE)
+
+	echo "File $($(1)_CACHE_DIR)/$(MOD_CACHE_FILE) saved in cache " >> $($(1)_DST_PATH)/$(1).log
+	echo "[ CACHE::SAVED ] $($(1)_CACHE_DIR)/$(MOD_CACHE_FILE)" >> $($(1)_DST_PATH)/$(1).log
+
+	$(call MOD_UNLOCK,$(1))
```

Comment:
What will happen if the MOD_UNLOCK not be called, if will block forever or not? Will it be unlocked automatically when the build process existed?

Reply:
When build process is exited/terminated, the lock file is automatically unlocked